File: data_mcp/gpt-5/go/rep9/main.go
package main

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"log"
	"net/http"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Data models

type User struct {
	ID       int    `json:"id"`
	Username string `json:"username"`
}

type internalUser struct {
	User
	Password string
}

type Todo struct {
	ID          int    `json:"id"`
	Title       string `json:"title"`
	Description string `json:"description"`
	Completed   bool   `json:"completed"`
	CreatedAt   string `json:"created_at"`
	UpdatedAt   string `json:"updated_at"`
	UserID      int    `json:"-"`
}

type Server struct {
	mu            sync.RWMutex
	usersByID     map[int]*internalUser
	usersByName   map[string]*internalUser
	nextUserID    int
	sessions      map[string]int // session token -> userID
	todosByID     map[int]*Todo
	nextTodoID    int
}

func NewServer() *Server {
	return &Server{
		usersByID:   make(map[int]*internalUser),
		usersByName: make(map[string]*internalUser),
		sessions:    make(map[string]int),
		todosByID:   make(map[int]*Todo),
		nextUserID:  1,
		nextTodoID:  1,
	}
}

// Utilities

func nowISO8601() string {
	// second precision UTC
	return time.Now().UTC().Format("2006-01-02T15:04:05Z")
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	if v == nil {
		return
	}
	enc := json.NewEncoder(w)
	enc.SetEscapeHTML(false)
	_ = enc.Encode(v)
}

func writeError(w http.ResponseWriter, status int, msg string) {
	writeJSON(w, status, map[string]string{"error": msg})
}

func (s *Server) generateToken() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

func (s *Server) getSessionUserID(r *http.Request) (int, bool) {
	cookie, err := r.Cookie("session_id")
	if err != nil || cookie.Value == "" {
		return 0, false
	}
	s.mu.RLock()
	defer s.mu.RUnlock()
	uid, ok := s.sessions[cookie.Value]
	return uid, ok
}

func (s *Server) requireAuth(next func(http.ResponseWriter, *http.Request, *internalUser)) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		uid, ok := s.getSessionUserID(r)
		if !ok {
			writeError(w, http.StatusUnauthorized, "Authentication required")
			return
		}
		s.mu.RLock()
		user := s.usersByID[uid]
		s.mu.RUnlock()
		if user == nil {
			writeError(w, http.StatusUnauthorized, "Authentication required")
			return
		}
		next(w, r, user)
	}
}

// Handlers

func (s *Server) handleRegister(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	defer r.Body.Close()
	var input struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	username := strings.TrimSpace(input.Username)
	password := input.Password

	if !validUsername(username) {
		writeError(w, http.StatusBadRequest, "Invalid username")
		return
	}
	if len(password) < 8 {
		writeError(w, http.StatusBadRequest, "Password too short")
		return
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	if _, exists := s.usersByName[username]; exists {
		writeError(w, http.StatusConflict, "Username already exists")
		return
	}
	id := s.nextUserID
	s.nextUserID++
	user := &internalUser{User: User{ID: id, Username: username}, Password: password}
	s.usersByID[id] = user
	s.usersByName[username] = user

	writeJSON(w, http.StatusCreated, user.User)
}

func (s *Server) handleLogin(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	defer r.Body.Close()
	var input struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}

	s.mu.RLock()
	user := s.usersByName[input.Username]
	s.mu.RUnlock()
	if user == nil || user.Password != input.Password {
		writeError(w, http.StatusUnauthorized, "Invalid credentials")
		return
	}
	tok, err := s.generateToken()
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Internal error")
		return
	}
	s.mu.Lock()
	s.sessions[tok] = user.ID
	s.mu.Unlock()

	cookie := &http.Cookie{Name: "session_id", Value: tok, Path: "/", HttpOnly: true}
	http.SetCookie(w, cookie)
	writeJSON(w, http.StatusOK, user.User)
}

func (s *Server) handleLogout(w http.ResponseWriter, r *http.Request, user *internalUser) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	cookie, err := r.Cookie("session_id")
	if err == nil && cookie.Value != "" {
		s.mu.Lock()
		delete(s.sessions, cookie.Value)
		s.mu.Unlock()
	}
	writeJSON(w, http.StatusOK, map[string]any{})
}

func (s *Server) handleMe(w http.ResponseWriter, r *http.Request, user *internalUser) {
	if r.Method != http.MethodGet {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	writeJSON(w, http.StatusOK, user.User)
}

func (s *Server) handlePassword(w http.ResponseWriter, r *http.Request, user *internalUser) {
	if r.Method != http.MethodPut {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	defer r.Body.Close()
	var input struct {
		Old string `json:"old_password"`
		New string `json:"new_password"`
	}
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	if input.Old != user.Password {
		writeError(w, http.StatusUnauthorized, "Invalid credentials")
		return
	}
	if len(input.New) < 8 {
		writeError(w, http.StatusBadRequest, "Password too short")
		return
	}
	s.mu.Lock()
	user.Password = input.New
	s.mu.Unlock()
	writeJSON(w, http.StatusOK, map[string]any{})
}

func (s *Server) handleTodos(w http.ResponseWriter, r *http.Request, user *internalUser) {
	switch r.Method {
	case http.MethodGet:
		// List
		s.mu.RLock()
		var list []*Todo
		for _, t := range s.todosByID {
			if t.UserID == user.ID {
				copy := *t
				list = append(list, &copy)
			}
		}
		s.mu.RUnlock()
		sort.Slice(list, func(i, j int) bool { return list[i].ID < list[j].ID })
		writeJSON(w, http.StatusOK, list)
	case http.MethodPost:
		defer r.Body.Close()
		var input struct {
			Title       *string `json:"title"`
			Description *string `json:"description"`
		}
		if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
			writeError(w, http.StatusBadRequest, "Invalid JSON")
			return
		}
		if input.Title == nil || strings.TrimSpace(*input.Title) == "" {
			writeError(w, http.StatusBadRequest, "Title is required")
			return
		}
		desc := ""
		if input.Description != nil {
			desc = *input.Description
		}
		now := nowISO8601()
		s.mu.Lock()
		id := s.nextTodoID
		s.nextTodoID++
		t := &Todo{ID: id, Title: strings.TrimSpace(*input.Title), Description: desc, Completed: false, CreatedAt: now, UpdatedAt: now, UserID: user.ID}
		s.todosByID[id] = t
		s.mu.Unlock()
		writeJSON(w, http.StatusCreated, t)
	default:
		writeError(w, http.StatusNotFound, "Not found")
	}
}

func (s *Server) parseTodoIDFromPath(path string) (int, error) {
	parts := strings.Split(strings.TrimPrefix(path, "/"), "/")
	if len(parts) < 2 || parts[0] != "todos" {
		return 0, errors.New("bad path")
	}
	id, err := strconv.Atoi(parts[1])
	if err != nil || id <= 0 {
		return 0, errors.New("bad id")
	}
	return id, nil
}

func (s *Server) handleTodoByID(w http.ResponseWriter, r *http.Request, user *internalUser) {
	id, err := s.parseTodoIDFromPath(r.URL.Path)
	if err != nil {
		writeError(w, http.StatusNotFound, "Todo not found")
		return
	}
	s.mu.RLock()
	t := s.todosByID[id]
	s.mu.RUnlock()
	if t == nil || t.UserID != user.ID {
		writeError(w, http.StatusNotFound, "Todo not found")
		return
	}
	switch r.Method {
	case http.MethodGet:
		writeJSON(w, http.StatusOK, t)
	case http.MethodPut:
		defer r.Body.Close()
		var input struct {
			Title       *string `json:"title"`
			Description *string `json:"description"`
			Completed   *bool   `json:"completed"`
		}
		if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
			writeError(w, http.StatusBadRequest, "Invalid JSON")
			return
		}
		if input.Title != nil && strings.TrimSpace(*input.Title) == "" {
			writeError(w, http.StatusBadRequest, "Title is required")
			return
		}
		s.mu.Lock()
		if input.Title != nil {
			t.Title = strings.TrimSpace(*input.Title)
		}
		if input.Description != nil {
			t.Description = *input.Description
		}
		if input.Completed != nil {
			t.Completed = *input.Completed
		}
		t.UpdatedAt = nowISO8601()
		s.mu.Unlock()
		writeJSON(w, http.StatusOK, t)
	case http.MethodDelete:
		s.mu.Lock()
		delete(s.todosByID, id)
		s.mu.Unlock()
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusNoContent)
	default:
		writeError(w, http.StatusNotFound, "Not found")
	}
}

var usernameRe = regexp.MustCompile(`^[a-zA-Z0-9_]{3,50}$`)

func validUsername(u string) bool { return usernameRe.MatchString(u) }

func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	// Ensure JSON content type on all responses except 204 handled in delete
	// We'll set default here; specific helpers also set it.
	w.Header().Set("Content-Type", "application/json")

	path := r.URL.Path
	switch {
	case r.Method == http.MethodPost && path == "/register":
		s.handleRegister(w, r)
	case r.Method == http.MethodPost && path == "/login":
		s.handleLogin(w, r)
	case path == "/logout":
		s.requireAuth(s.handleLogout)(w, r)
	case path == "/me":
		s.requireAuth(s.handleMe)(w, r)
	case path == "/password":
		s.requireAuth(s.handlePassword)(w, r)
	case path == "/todos" || path == "/todos/":
		s.requireAuth(s.handleTodos)(w, r)
	case strings.HasPrefix(path, "/todos/"):
		s.requireAuth(s.handleTodoByID)(w, r)
	default:
		writeError(w, http.StatusNotFound, "Not found")
	}
}

func main() {
	var port int
	flag.IntVar(&port, "port", 8080, "port to listen on")
	flag.Parse()

	addr := fmt.Sprintf("0.0.0.0:%d", port)
	srv := &http.Server{Addr: addr, Handler: NewServer()}
	log.Printf("listening on %s", addr)
	if err := srv.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
		log.Fatalf("server error: %v", err)
	}
}

File: data_mcp/gpt-5/java/rep9/src/Main.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;
import com.google.gson.*;

import java.io.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.regex.Pattern;

public class Main {
    // Data models
    static class User {
        final int id;
        final String username;
        String password; // stored as plain for simplicity in this in-memory demo
        User(int id, String username, String password) {
            this.id = id; this.username = username; this.password = password;
        }
    }
    static class Todo {
        final int id;
        final int userId;
        String title;
        String description;
        boolean completed;
        String created_at;
        String updated_at;
        Todo(int id, int userId, String title, String description, boolean completed, String ts) {
            this.id = id; this.userId = userId; this.title = title; this.description = description;
            this.completed = completed; this.created_at = ts; this.updated_at = ts;
        }
    }

    static class State {
        final ConcurrentHashMap<Integer, User> usersById = new ConcurrentHashMap<>();
        final ConcurrentHashMap<String, Integer> userIdByUsername = new ConcurrentHashMap<>();
        final ConcurrentHashMap<String, Integer> sessions = new ConcurrentHashMap<>();
        final ConcurrentHashMap<Integer, Todo> todosById = new ConcurrentHashMap<>();
        final AtomicInteger userIdSeq = new AtomicInteger(0);
        final AtomicInteger todoIdSeq = new AtomicInteger(0);
    }

    static final Gson gson = new GsonBuilder().disableHtmlEscaping().create();
    static final Pattern USERNAME_RE = Pattern.compile("^[a-zA-Z0-9_]{3,50}$");

    public static void main(String[] args) throws Exception {
        int port = 8080;
        for (int i = 0; i < args.length; i++) {
            if ("--port".equals(args[i]) && i + 1 < args.length) {
                port = Integer.parseInt(args[i+1]);
                i++;
            }
        }
        State state = new State();
        HttpServer server = HttpServer.create(new InetSocketAddress("0.0.0.0", port), 0);
        server.createContext("/", new Router(state));
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("Server listening on 0.0.0.0:" + port);
    }

    static class Router implements HttpHandler {
        private final State state;
        Router(State state) { this.state = state; }
        @Override public void handle(HttpExchange ex) throws IOException {
            try {
                String method = ex.getRequestMethod();
                URI uri = ex.getRequestURI();
                String path = uri.getPath();

                // Normalize: remove trailing slash except root
                if (path.length() > 1 && path.endsWith("/")) {
                    path = path.substring(0, path.length()-1);
                }

                if (method.equals("POST") && path.equals("/register")) { handleRegister(ex, state); return; }
                if (method.equals("POST") && path.equals("/login")) { handleLogin(ex, state); return; }
                if (method.equals("POST") && path.equals("/logout")) { handleLogout(ex, state); return; }
                if (method.equals("GET") && path.equals("/me")) { handleMe(ex, state); return; }
                if (method.equals("PUT") && path.equals("/password")) { handlePassword(ex, state); return; }
                if (path.equals("/todos")) {
                    if (method.equals("GET")) { handleTodosList(ex, state); return; }
                    if (method.equals("POST")) { handleTodosCreate(ex, state); return; }
                }
                if (path.startsWith("/todos/")) {
                    String idStr = path.substring("/todos/".length());
                    Integer id = parseId(idStr);
                    if (id == null) {
                        sendError(ex, 404, "Not found");
                        return;
                    }
                    if (method.equals("GET")) { handleTodosGet(ex, state, id); return; }
                    if (method.equals("PUT")) { handleTodosUpdate(ex, state, id); return; }
                    if (method.equals("DELETE")) { handleTodosDelete(ex, state, id); return; }
                }
                sendError(ex, 404, "Not found");
            } catch (Exception e) {
                // Log and return 500
                e.printStackTrace();
                sendError(ex, 500, "Internal server error");
            } finally {
                ex.close();
            }
        }
    }

    static void handleRegister(HttpExchange ex, State state) throws IOException {
        JsonObject body = parseJsonObject(ex);
        if (body == null) { sendError(ex, 400, "Invalid JSON"); return; }
        String username = getAsString(body, "username");
        String password = getAsString(body, "password");
        if (username == null || !USERNAME_RE.matcher(username).matches()) {
            sendError(ex, 400, "Invalid username"); return;
        }
        if (password == null || password.length() < 8) {
            sendError(ex, 400, "Password too short"); return;
        }
        // Uniqueness
        if (state.userIdByUsername.putIfAbsent(username, -1) != null) { // temp mark to avoid race
            sendError(ex, 409, "Username already exists");
            return;
        }
        try {
            int id = state.userIdSeq.incrementAndGet();
            User user = new User(id, username, password);
            state.usersById.put(id, user);
            state.userIdByUsername.put(username, id);
            JsonObject resp = new JsonObject();
            resp.addProperty("id", id);
            resp.addProperty("username", username);
            sendJson(ex, 201, resp);
        } finally {
            // nothing to cleanup
        }
    }

    static void handleLogin(HttpExchange ex, State state) throws IOException {
        JsonObject body = parseJsonObject(ex);
        if (body == null) { sendError(ex, 400, "Invalid JSON"); return; }
        String username = getAsString(body, "username");
        String password = getAsString(body, "password");
        if (username == null || password == null) {
            sendError(ex, 401, "Invalid credentials"); return;
        }
        Integer uid = state.userIdByUsername.get(username);
        if (uid == null || uid <= 0) {
            sendError(ex, 401, "Invalid credentials"); return;
        }
        User u = state.usersById.get(uid);
        if (u == null || !u.password.equals(password)) {
            sendError(ex, 401, "Invalid credentials"); return;
        }
        String token = UUID.randomUUID().toString().replace("-", "");
        state.sessions.put(token, u.id);
        ex.getResponseHeaders().add("Set-Cookie", "session_id=" + token + "; Path=/; HttpOnly");
        JsonObject resp = new JsonObject();
        resp.addProperty("id", u.id);
        resp.addProperty("username", u.username);
        sendJson(ex, 200, resp);
    }

    static void handleLogout(HttpExchange ex, State state) throws IOException {
        User u = requireAuth(ex, state);
        if (u == null) return; // response already sent
        String token = getSessionToken(ex);
        if (token != null) {
            state.sessions.remove(token);
        }
        // Return empty object per spec
        JsonObject obj = new JsonObject();
        sendJson(ex, 200, obj);
    }

    static void handleMe(HttpExchange ex, State state) throws IOException {
        User u = requireAuth(ex, state);
        if (u == null) return;
        JsonObject resp = new JsonObject();
        resp.addProperty("id", u.id);
        resp.addProperty("username", u.username);
        sendJson(ex, 200, resp);
    }

    static void handlePassword(HttpExchange ex, State state) throws IOException {
        User u = requireAuth(ex, state);
        if (u == null) return;
        JsonObject body = parseJsonObject(ex);
        if (body == null) { sendError(ex, 400, "Invalid JSON"); return; }
        String oldp = getAsString(body, "old_password");
        String newp = getAsString(body, "new_password");
        if (oldp == null || !u.password.equals(oldp)) {
            sendError(ex, 401, "Invalid credentials"); return;
        }
        if (newp == null || newp.length() < 8) {
            sendError(ex, 400, "Password too short"); return;
        }
        u.password = newp;
        sendJson(ex, 200, new JsonObject());
    }

    static void handleTodosList(HttpExchange ex, State state) throws IOException {
        User u = requireAuth(ex, state);
        if (u == null) return;
        List<Todo> list = new ArrayList<>();
        for (Todo t : state.todosById.values()) {
            if (t.userId == u.id) list.add(t);
        }
        list.sort(Comparator.comparingInt(a -> a.id));
        JsonArray arr = new JsonArray();
        for (Todo t : list) arr.add(todoToJson(t));
        sendJson(ex, 200, arr);
    }

    static void handleTodosCreate(HttpExchange ex, State state) throws IOException {
        User u = requireAuth(ex, state);
        if (u == null) return;
        JsonObject body = parseJsonObject(ex);
        if (body == null) { sendError(ex, 400, "Invalid JSON"); return; }
        String title = getAsString(body, "title");
        if (title == null || title.trim().isEmpty()) {
            sendError(ex, 400, "Title is required"); return;
        }
        String description = getAsString(body, "description");
        if (description == null) description = "";
        int id = state.todoIdSeq.incrementAndGet();
        String ts = nowIso();
        Todo t = new Todo(id, u.id, title, description, false, ts);
        state.todosById.put(id, t);
        sendJson(ex, 201, todoToJson(t));
    }

    static void handleTodosGet(HttpExchange ex, State state, int id) throws IOException {
        User u = requireAuth(ex, state);
        if (u == null) return;
        Todo t = state.todosById.get(id);
        if (t == null || t.userId != u.id) {
            sendError(ex, 404, "Todo not found"); return;
        }
        sendJson(ex, 200, todoToJson(t));
    }

    static void handleTodosUpdate(HttpExchange ex, State state, int id) throws IOException {
        User u = requireAuth(ex, state);
        if (u == null) return;
        Todo t = state.todosById.get(id);
        if (t == null || t.userId != u.id) {
            sendError(ex, 404, "Todo not found"); return;
        }
        JsonObject body = parseJsonObject(ex);
        if (body == null) { sendError(ex, 400, "Invalid JSON"); return; }
        if (body.has("title") && !body.get("title").isJsonNull()) {
            String title = getAsString(body, "title");
            if (title == null || title.trim().isEmpty()) {
                sendError(ex, 400, "Title is required"); return;
            }
            t.title = title;
        }
        if (body.has("description")) {
            String desc = null;
            if (!body.get("description").isJsonNull()) desc = getAsString(body, "description");
            t.description = desc == null ? "" : desc;
        }
        if (body.has("completed") && !body.get("completed").isJsonNull()) {
            JsonElement ce = body.get("completed");
            if (ce.isJsonPrimitive() && ((JsonPrimitive)ce).isBoolean()) {
                t.completed = ce.getAsBoolean();
            } else {
                // if provided but not boolean, treat as error? Not specified; ignore or error. We'll coerce error 400.
                sendError(ex, 400, "Invalid JSON"); return;
            }
        }
        t.updated_at = nowIso();
        sendJson(ex, 200, todoToJson(t));
    }

    static void handleTodosDelete(HttpExchange ex, State state, int id) throws IOException {
        User u = requireAuth(ex, state);
        if (u == null) return;
        Todo t = state.todosById.get(id);
        if (t == null || t.userId != u.id) {
            sendError(ex, 404, "Todo not found"); return;
        }
        state.todosById.remove(id);
        // DELETE should return 204 no body and no Content-Type per spec
        ex.getResponseHeaders().remove("Content-Type");
        ex.sendResponseHeaders(204, -1);
    }

    // Helpers
    static String nowIso() {
        return Instant.now().truncatedTo(ChronoUnit.SECONDS).toString();
    }

    static Integer parseId(String s) {
        try { return Integer.parseInt(s); } catch (Exception e) { return null; }
    }

    static JsonObject todoToJson(Todo t) {
        JsonObject o = new JsonObject();
        o.addProperty("id", t.id);
        o.addProperty("title", t.title);
        o.addProperty("description", t.description);
        o.addProperty("completed", t.completed);
        o.addProperty("created_at", t.created_at);
        o.addProperty("updated_at", t.updated_at);
        return o;
    }

    static JsonObject parseJsonObject(HttpExchange ex) throws IOException {
        String body = readBody(ex);
        if (body == null) return null;
        try {
            JsonElement el = JsonParser.parseString(body);
            if (!el.isJsonObject()) return null;
            return el.getAsJsonObject();
        } catch (JsonSyntaxException e) {
            return null;
        }
    }

    static String readBody(HttpExchange ex) throws IOException {
        InputStream is = ex.getRequestBody();
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        byte[] buf = new byte[4096];
        int r;
        while ((r = is.read(buf)) != -1) baos.write(buf, 0, r);
        return baos.toString(StandardCharsets.UTF_8);
    }

    static String getAsString(JsonObject obj, String key) {
        if (!obj.has(key)) return null;
        JsonElement el = obj.get(key);
        if (el == null || el.isJsonNull()) return null;
        if (!el.isJsonPrimitive()) return null;
        return el.getAsString();
    }

    static void sendJson(HttpExchange ex, int status, JsonElement el) throws IOException {
        byte[] bytes = gson.toJson(el).getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().set("Content-Type", "application/json");
        ex.sendResponseHeaders(status, bytes.length);
        try (OutputStream os = ex.getResponseBody()) {
            os.write(bytes);
        }
    }
    static void sendJson(HttpExchange ex, int status, JsonObject obj) throws IOException {
        sendJson(ex, status, (JsonElement) obj);
    }
    static void sendJson(HttpExchange ex, int status, JsonArray obj) throws IOException {
        sendJson(ex, status, (JsonElement) obj);
    }

    static void sendError(HttpExchange ex, int status, String message) throws IOException {
        JsonObject err = new JsonObject();
        err.addProperty("error", message);
        sendJson(ex, status, err);
    }

    static String getSessionToken(HttpExchange ex) {
        List<String> cookies = ex.getRequestHeaders().get("Cookie");
        if (cookies == null) return null;
        for (String header : cookies) {
            String[] parts = header.split(";\\s*");
            for (String part : parts) {
                int eq = part.indexOf('=');
                if (eq <= 0) continue;
                String name = part.substring(0, eq).trim();
                String value = part.substring(eq+1).trim();
                if (name.equals("session_id")) return value;
            }
        }
        return null;
    }

    static User requireAuth(HttpExchange ex, State state) throws IOException {
        String token = getSessionToken(ex);
        if (token == null) { sendError(ex, 401, "Authentication required"); return null; }
        Integer uid = state.sessions.get(token);
        if (uid == null) { sendError(ex, 401, "Authentication required"); return null; }
        User u = state.usersById.get(uid);
        if (u == null) { sendError(ex, 401, "Authentication required"); return null; }
        return u;
    }
}
